feat: password change email

zlwaterfield opened the pull request:
> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Changes

We don't have an email to the user when their password is changed, this adds one.

- add new customer.io template
- add django template 
- add task to send email
- trigger task when password is changed

email
<img width="814" height="722" alt="Screenshot 2025-07-10 at 12 17 35 PM" src="https://github.com/user-attachments/assets/0369ef6a-6e60-46e4-866d-2575fd237d1b" />

Other fixes
- API error `attr` fix
- change password erroring rendering fix 

## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->

- [ ] I've [added or updated the docs](https://posthog.com/handbook/engineering/writing-docs)
- [ ] I've reached out for help from the docs team
- [ ] No docs needed for this change

## How did you test this code?

Manually

zlwaterfield committed:
add attr support for api errors

zlwaterfield committed:
fix password validation empty string issue

zlwaterfield committed:
add a timeout so error renders

zlwaterfield committed:
shift up out of password changes

zlwaterfield committed:
add password changed email

zlwaterfield committed:
LINT

zlwaterfield committed:
Add email notification test for password changes

Tests for user password updates now assert that the password changed email is sent. Added mocking for send_password_changed_email.delay in relevant test cases to verify email notification behavior.

zlwaterfield committed:
Update posthog/tasks/email.py

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

zlwaterfield committed:
make sync

joshsny commented on posthog/tasks/email.py:
```diff
@@ -183,6 +183,24 @@ def send_password_reset(user_id: int, token: str) -> None:
+        },
+    )
+    message.add_recipient(user.email)
+    message.send(send_async=False)
```
do we want it sync?

joshsny commented on posthog/tasks/email.py:
```diff
@@ -183,6 +183,24 @@ def send_password_reset(user_id: int, token: str) -> None:
+    user = User.objects.get(pk=user_id)
+    message = EmailMessage(
+        use_http=True,
+        campaign_key=f"password-changed-{user.uuid}-{timezone.now().timestamp()}",
```
how does this work with retries, not sure what our default is set to in celery but if the task fails and retries this campaign key wouldn't prevent it from sending two emails

zlwaterfield replied:
This is simply for idepempotency, it's not relevant for emails like this, typically more ones that could be triggered multiple times, but we only want to send it once, in that case we wouldn't put a timestamp so it's distinct to the event.

joshsny reviewed (approved):
LGTM

zlwaterfield commented on posthog/templates/email/password_changed.html:
```diff
@@ -0,0 +1,6 @@
+{% extends "email/base.html" %} {% load posthog_assets %} {% load posthog_filters %}{% block section %}
```
1. Django templates are used as fallbacks so for open source when customer.io is not configured or when we don't have a template for it in customer.io - so they shouldn't be sent for cloud. We have a section in the docs about this: https://posthog.com/handbook/engineering/developing-locally#extra-sending-emails
2. Where do you see IP, device, etc. on the django template - we don't currently use this as all in the customer.io templates but it's something I'd like to revisit and add this quarter
3. Will do 👍

zlwaterfield committed:
Send password changed email asynchronously

Removed the send_async=False argument to allow the password changed email to be sent asynchronously, aligning with the default behavior of the message.send() method.
